programs/host/frontend.c

```diff
@@ -0,0 +1,943 @@
+{
+    int i, file_id = -1;
+
+    for ( i = 0; i < MBEDTLS_SERIALIZE_MAX_FILES; i++ )
+    {
+        if ( !files[i].inUse )
+        {
+            files[i].inUse = 1;
+            file_id = i + 1; // file Id can not be 0 i.e. same as NULL
+            break;
+        }
+    }
+    return file_id;
+}
+
+int free_file_context( int32_t file_id )
+{
+    int ret = -1;
+    file_id--;
+    if ( file_id < MBEDTLS_SERIALIZE_MAX_FILES )
+    {
+        files[file_id].inUse = 0;
+        if ( files[file_id].file )
+        {
+            files[file_id].file = NULL;
+            ret = 0;
+        }
+    }
+    return ret;
+}
+
+void * get_file_from_id( int32_t file_id )
+{
+    file_id--;
+    if ( file_id < MBEDTLS_SERIALIZE_MAX_FILES && files[file_id].inUse )
+    {
+        return files[file_id].file;
+    }
+    DBG( "Failed to find file for Index %d", file_id );
+    return NULL;
+}
+
+/** Execute an offloaded function.
+ *
+ * @param ctx       Offloading contecxt
+ * @param function  Function to execute (MBEDTLS_SERIALIZE_FUNCTION_xxx)
+ * @param outputs   Array[16] of outputs. Initially all-null. On return,
+ *                  for a function with N outputs, positions 0 to N-1 will
+ *                  be filled with output parameters allocated with alloc_item.
+ * @return status (0 for ok, MBEDTLS_ERR_xxx on error)
+ */
+static uint32_t mbedtls_serialize_perform( mbedtls_serialize_context_t *ctx_p,
+                                           uint32_t function,
+                                           mbedtls_serialize_item_t **outputs )
+{
+    mbedtls_serialize_item_t *inputs[16] = {0};
+    mbedtls_serialize_item_t *item;
+    size_t arity;
+    int ret = 0;
+
+    for( arity = 0, item = ctx_p->stack;
+         arity < ( ( function & 0x0000f0 ) >> 4 ) && item != NULL;
+         arity++, item = item->next )
+    {
+        inputs[arity] = item;
+    }
+    DBG( "arity=%zu", arity );
+
+    switch( function )
+    {
+    case MBEDTLS_SERIALIZE_FUNCTION_EXIT:
+        ret = 0;
+        ctx_p->status = MBEDTLS_SERIALIZE_STATUS_EXITED;
+        break;
+
+    case MBEDTLS_SERIALIZE_FUNCTION_ECHO:
+    {
+        CHECK_ARITY( 1 );
+        ALLOC_OUTPUT( 0, inputs[0]->size );
+        DBG( "executing echo" );
+        memcpy( item_buffer( outputs[0] ), item_buffer( inputs[0] ), inputs[0]->size );
+        ret = 0;
+        break;
+    }
+
+    case MBEDTLS_SERIALIZE_FUNCTION_USLEEP:
+        CHECK_ARITY( 1 );
+        CHECK_LENGTH( 0, 4 ); // usec
+        {
+            unsigned long usec = item_int32( inputs[0] );
+            DBG( "executing sleep usec=%lu", usec );
+            mbedtls_net_usleep( usec );
+            ret = 0;
+        }
+        break;
+
+    case MBEDTLS_SERIALIZE_FUNCTION_SOCKET:
+        CHECK_ARITY( 3 ); // host, port, proto_and_mode
+        CHECK_LENGTH( 2, 2 ); // proto_and_mode
+        {
+            char *host = item_buffer( inputs[0] );
+            char *port = item_buffer( inputs[1] );
```

Comment:
For safety, we could insert a check here that `host` and `port` are properly null-terminated.